### ISSUE TITLE
Incorrect Handling of WP Subdirectory Install on Legacy Multisite

### ISSUE THREAD

**Issue by elyobo** [deliciousbrains/wp-amazon-s3-and-cloudfront#255]
I have a multisite installation with wordpress installed in a wp/ subdirectory.  The installation was upgraded from an old WP version (pre WP 3.5).

The plugin ends pushing files up to S3 with the full directory structure of the site (e.g. /var/www/whatever) as part of the S3 URL.  This looks to be due to differences between wp_upload_dir(), which has a lot of logic specific to multisite upload path handling, including backwards compatible fixes for pre 3.5 versions, and Amazon_S3_And_CloudFront::get_base_upload_path() which replicates only a small part of this logic.  The base path calculated is therefore incorrect, causing the prefix generation to fail (it attempts to strip the incorrect base upload path from the wp_upload_dir path).

**Comment by elyobo** [comment]
Fix as suggested above in #256.

**Comment by polevaultweb** [comment]
@elyobo could you post some of your MS and custom upload constants from `wp-config.php` so I can try to recreate?

What version of MS was this originally? Has it been upgraded to the latest WP 4.2.3?

**Comment by elyobo** [comment]
@polevaultweb thanks for your response.

The original version was ancient, I'm not sure what; it has recently been upgraded from a reasonably low number 3.x series to 4.2.3, just to keep it ticking over.

It doesn't have a custom UPLOADS constant, but when a new multisite blog is created an upload_path is set, which triggers the problem.  The upload_path option is set to "wp-content/blogs.dir/2/files", which is relative to my installation root (/var/www/whatever) not where my WP is (/var/www/whatever/wp), i.e. the uploads go into /var/www/whatever/wp-content/blogs.dir/2/files, not into /var/www/whatever/wp/wp-content/blogs.dir/2/files.

That should be sufficient to trigger the problem

* Install WP as multisite in a subdirectory
* Create a new blog (this may already be sufficient; I'm not sure what the newer versions put in the upload_path value)
* If it's not occurring already, try setting the upload_path option to my value from above, "wp-content/blogs.dir/2/files"

**Comment by elyobo** [comment]
Creating a new blog in the same multisite installation suffers from the same problem; the upload_path is set in the same way (e.g. wp-content/blogs.dir/my-blog-id/files).

I also notice that the documentation for [wp_upload_dir](https://core.trac.wordpress.org/browser/tags/4.2.2/src/wp-includes/functions.php#L1693) (which your code seems to be heavily based on) is a little inconsistent with the implementation.  It says in the docblock that it "[c]hecks the 'upload_path' option, which should be from the web root folder" however in the code it does the same as yours and concats the upload_path with the ABSPATH (which is not the web root folder in a subdirectory install).

Changing the upload_path so that it was relative to the ABSPATH (as [UPLOADS](https://codex.wordpress.org/Determining_Plugin_and_Content_Directories#Constants) is apparently meant to be, but which contradicts the comments on wp_content_dir, didn't fix things either, as the relative portions don't match those returned by wp_content_dir, so the str_replace still fails.  However if you wrap the result of get_base_upload_path() in realpath, *and* set an upload_path relative to ABSPATH (e.g. "../wp-content/blogs.dir/2/files") that does also work.

**Comment by elyobo** [comment]
After digging around more, this only seems to affect multisite installs that have been upgraded from pre-3.5 multisite.  A recent install of multisite is fine (doesn't set upload_path), and recent non-multisite is fine even when an upload_path or UPLOADS constant is set.

So this appears to affect only multisite installations, with WP in a subdirectory, that have been upgraded from an old WP version.

**Comment by elyobo** [comment]
@polevaultweb It looks like the problem is due to differences between your base upload plugin path detection (which doesn't have any extra logic for legay multisite handling) vs wp_upload_dir (which does).  I've cancelled my PR, which seems inappropriate now, and updated the original description.

A less than ideal but still helpful fix would probably be to add a filter that allows us to hook in and modify the prefix when it's generated; this would be safely backwards compatible but still allow those of us with unusual situations to fix things up so that they work correctly.

Other solutions would be to more fully replicate the logic of wp_upload_dir (which is possibly fragile, as you need to keep up to date with any changes to it - but that is the situation you're already in, I guess) or to throw away get_base_upload_path entirely and just strip off the year month folders from the end if they're there (might fail in some corner cases where someone is applying a filter on upload_dir).

**Comment by elyobo** [comment]
Along those lines, I have been able to work around this issue by adding my own filter to upload_dir to modify the 'path' field so that it contains the same structure as is returned by Amazon_S3_And_CloudFront::get_base_upload_path().

Code is something along the lines of the following.  It should disable itself automatically if the backwards compatibility issue is fixed.  Note that, because the upload directory is changes, PHP must have write access to the directory that AS3F expects (i.e. within subdir/wp-content/blogs.dir in my case).

```php
        if (!is_multisite() || !defined('BLOGUPLOADDIR')) {
            // Fix does not apply
            return;
        }

        if (!is_plugin_active('amazon-s3-and-cloudfront/wordpress-s3.php')) {
            // Fix does not apply
            return;
        }

        add_filter('upload_dir', function ($up) {
            // Only apply if we have used the BLOGUPLOADDIR
            if ($up['basedir'] != rtrim(BLOGUPLOADDIR, '/')) {
                return $up;
            }

            // Strict standards: this is not actually static, but the method
            // is stateless so it's OK to call
            $s3base = @\Amazon_S3_And_CloudFront::get_base_upload_path();

            if (false !== strpos($s3base, BLOGUPLOADDIR)) {
                // The upload dir is already in there, so the bug has been fixed
                return $up;
            }

            // Replace the path with one that will work with AS3F
            $suffix     = str_replace($up['basedir'], null, $up['path']);
            $up['path'] = rtrim($s3base, '/') . '/' . ltrim($suffix, '/');

            return $up;
        });
```

**Comment by A5hleyRich** [comment]
Resolved in version 0.9.

**Comment by elyobo** [comment]
Thanks @A5hleyRich, I'll check it out.
